python/lsst/cp/pipe/defects.py

```diff
@@ -0,0 +1,645 @@
+        nBad = countMaskedPixels(maskedIm, badMaskString)
+        return nPixels - nBad
+
+    def findHotAndColdPixels(self, exp, imageType, setMask=False):
+        """Find hot and cold pixels in an image.
+
+        Using config-defined thresholds on a per-amp basis, mask pixels
+        that are nSigma above threshold in dark frames (hot pixels),
+        or nSigma away from the clipped mean in flats (hot & cold pixels).
+
+        Parameters
+        ----------
+        setMask: `bool`
+            If true, update exp with hot and cold pixels
+            hot: DETECTED
+            cold: DETECTED_NEGATIVE
+
+        Returns
+        -------
+        defects: `lsst.meas.algorithms.interp.Defect`
+            The defects
+        """
+        assert imageType in ['flat', 'dark']
+
+        self._setEdgeBits(exp)
+        maskedIm = exp.maskedImage
+
+        polarities = {'dark': [True], 'flat': [True, False]}[imageType]
+
+        footprintList = []
+
+        for amp in exp.getDetector():
+            ampImg = maskedIm[amp.getBBox()].clone()
+
+            # XXX can we skip this if we set the stats box to ignore the EDGE bit?
+            # crop ampImage depending on where the amp lies in the image
+            if self.config.nPixBorderLeftRight:
+                if ampImg.getX0() == 0:
+                    ampImg = ampImg[self.config.nPixBorderLeftRight:, :, afwImage.LOCAL]
+                else:
+                    ampImg = ampImg[:-self.config.nPixBorderLeftRight, :, afwImage.LOCAL]
+            if self.config.nPixBorderUpDown:
+                if ampImg.getY0() == 0:
+                    ampImg = ampImg[:, self.config.nPixBorderUpDown:, afwImage.LOCAL]
+                else:
+                    ampImg = ampImg[:, :-self.config.nPixBorderUpDown, afwImage.LOCAL]
+
+            if self._getNumGoodPixels(ampImg) == 0:  # amp contains no usable pixels
+                continue
+
+            # XXX needs better initialisation, and config options for the MEANCLIP
+            ampImg -= afwMath.makeStatistics(ampImg, afwMath.MEANCLIP, ).getValue()
+
+            mergedSet = None
+            for polarity in polarities:
+                nSig = self.config.nSigmaBright if polarity else self.config.nSigmaDark
+                threshold = afwDetection.createThreshold(nSig, 'stdev', polarity=polarity)
+
+                footprintSet = afwDetection.FootprintSet(ampImg, threshold)
+                if setMask:
+                    footprintSet.setMask(maskedIm.mask, ("DETECTED" if polarity else "DETECTED_NEGATIVE"))
+
+                if mergedSet is None:
+                    mergedSet = footprintSet
+                else:
+                    mergedSet.merge(footprintSet)
+
+            footprintList += mergedSet.getFootprints()
+
+        defects = measAlg.Defects.fromFootprintList(footprintList)
+        return defects
+
+    def _plot(self, dataRef, exp, visit, nSig, defects, imageType):  # pragma: no cover
```

Comment:
I don't think we have a policy, at least that I've seen. It's a tricky thing. I think the normal pattern is that plots are really only made as part of lsstDebug blocks, and are therefore untested and often even totally bitrotted. I want to change that for `cp_pipe` as it's a different MO there really, and I think plotting some thing should be default. However, test coverage can't really be much more than checking if the file exists I think, and that's _almost_ pointless (not quite, but close) and given that visual inspection inherently means a human eye is needed I think it's not worth the hassle. We can discuss if you disagree though.

Reply:
My general view is that you should write tests for the code that does the math that goes into the plots (which I think you mostly have here), but then I'm not sure how to test the actually plotting stuff itself. Since I don't have a good solution, I'm not going to fight about it (or the lack of coverage), but we should probably ask an expert (on #dm-square?).